### DISCUSSION TITLE
modbus communication

### DISCUSSION THREAD

**Discussion by sansanda** [pymeasure/pymeasure#751] [Q&A]
Hi All.

I'm actually very interested in controlling/communicating different equipment (like temperature controllers, sensors, etc.) via the modbus rtu protocol over serial communication (i.e. RS-232 or RS-485). 

Modbus uses ID (plus the COM port of the serial comm) integer to address (ask, configure, etc.) the different devices over the serial network.
My doubt is, is it necessary or convenient to develop a new adapter in the pymeasure/adapters folder (extending pyserial)? or it would be better use only pyserial at instrument level and solve the modbus particularities adding an ID variable?

Many thanks,

David.

**Comment by bilderbuchi** [comment]
Can that communicate via pyvisa, too? Are the docs on the binary/frame-based instruments useful for what you want to achieve? https://pymeasure.readthedocs.io/en/latest/dev/adding_instruments.html#bytes-communication

**Reply by BenediktBurger** [reply]
If it is possible to easily write a code with pyvisa, it would be good to use it.
If that is possible, we could make a `ModbusInstrument`, which has those methods implemented, in order to use that for more instruments, as I understand Modbus is a standard and might be used with other instruments.

If it is a too large hassle to write the modbus code for pyvisa (instead of using pymodbus or something), we can talk again about a new adapter.

I looked at some documentation regarding modbus and it seems, that it is just some encoding of bytes and should be very easy to implement with pyvisa. (should...)

**Reply by bilderbuchi** [reply]
From what I remember, modbus is similar to the [TC038 instrument](https://github.com/pymeasure/pymeasure/blob/master/pymeasure/instruments/hcp/tc038d.py) in that you supply an operation (read/write), an address (or address range) and a value (or values). There are some peculiarities (a single address is called a "coil" iirc), and I don't know how much validation etc goes on in the background.

If @sansanda thought he could use pyserial, I don't see why we could not use pyvisa (and hence our default VISAAdapter), instead. However, see above. 
Using the existing adapter would be convenient for us, if the additional effort (concerning methods, logic, etc.) is low. However, I would not want to basically reimplement pymodbus or minimalmodbus. Maybe a `ModbusInstrument` could be a good solution, if there are a handful of common, simple, building blocks.

I don't know, it *could* be that it's similar to the PRologixAdapter, in that it's a simple protocol with some additional stuff around that - then a separate adapter (possibly wrapping a modbus library) could be the way to go (however: additional dependency).

Maybe a concrete example of some communication details would help. @sansanda do you have something you can show/explain?

**Reply by BenediktBurger** [reply]
Oh, yes, if I remember correctly the manual of the temperature controller built into the TC38D instrument talked about modbus. I never cared, what exactly that is, only that it followed my commands :grin:

**Reply by bilderbuchi** [reply]
I once had a delightful time interfacing to a modus driver for an industrial stepper motor with a 480(!) page manual. 😭

**Comment by mcdo0486** [comment]
> Hi All.
> 
> I'm actually very interested in controlling/communicating different equipment (like temperature controllers, sensors, etc.) via the modbus rtu protocol over serial communication (i.e. RS-232 or RS-485).
> 
> Modbus uses ID (plus the COM port of the serial comm) integer to address (ask, configure, etc.) the different devices over the serial network. My doubt is, is it necessary or convenient to develop a new adapter in the pymeasure/adapters folder (extending pyserial)? or it would be better use only pyserial at instrument level and solve the modbus particularities adding an ID variable?
> 
> Many thanks,
> 
> David.

I found that including the python library minimalmodbus very useful. It is a pure python implementation with a lot of helpful functions for reading and writing registers, longs, and floats
https://minimalmodbus.readthedocs.io/en/stable/readme.html

I included it as a dependency in my pymeasure fork to support the Omega CS8DPT temperature controller. 
https://github.com/mcdo0486/pymeasure/blob/6d5de8eaba35bb1d1b4ab98b91859b29d6def551/pymeasure/instruments/omega/adapters.py

I haven't pushed that branch yet because it adds a dependency to pymeasure, and I have to clean it up. But I think minimalmodbus should be included because it is used by a lot of instruments.

**Reply by BenediktBurger** [reply]
I had a look at minimummodbus, and yes, it offers some convenience by doing the error checks and Co. But some of them are not necessary, as `int.to_bytes` offers already some transformation. 
However, doing modbus is not that difficult. I just found out, that the hcp.tc038d device (see bilderbuchi's link) already implements modbus (the function codes are the same).
Therefore I prefer to not introduce that dependency.

**Reply by mcdo0486** [reply]
I'm not an expert on modbus protocol, but what I like about minimalmodbus is that it is one file (be it 4K lines) with really detailed inline documentation. It also supports ASCII mode with LRC checking, the hcp.tc038d implementation only supports RTU mode with CRC.  The project seems to have gone through experience dealing with [manufacture quirks](https://github.com/pyhys/minimalmodbus/blob/185afffd2d65b585f69bc5e3f640270d2bd27444/minimalmodbus.py#L448) in modbus implementation. That's why I brought it up in this discussion of pymeasure creating its own implementation of the modbus protocol.

https://github.com/pyhys/minimalmodbus/blob/master/minimalmodbus.py

**Reply by bilderbuchi** [reply]
There's also https://github.com/riptideio/pymodbus and probably a couple of other projects. Before choosing we also have to look at the health of the respective projects (performance, activity level, bug tracker, release cadence) -- it would suck to find a problem and not being able to reach someone for a fix.

It seems to me that a simple version of the protocol seems in reach for implementing ourselves. A more fully-fledged one, including all kinds of modes and intricacies could be more involved. 
It also depends on how much we have to change/shim/wrap to make the external package conform to our API, especially given the increased number of configurations/"modes" involved that we have to give a possibility of choosing. 
From the link to your custom adapter, it looks reasonable. In any case, this should be generic enough that we can switch libraries internally, if needed.
However, looking further, there's [a lot of boilerplate going on here](https://github.com/mcdo0486/pymeasure/blob/6d5de8eaba35bb1d1b4ab98b91859b29d6def551/pymeasure/instruments/omega/omega_registers.py).

@bmoneke also keep in mind, if we do a minimal version now to support modbus for the few instruments in question, I'm sure someone will come out of the woods with a modbus instrument that needs another set of modes/flags. :D

**Reply by bilderbuchi** [reply]
Another avenue would be a barebones self-made implementation, with a fallback (or for more complex use-cases) to an optional external package if it's installed (or with a config switch).

**Reply by mcdo0486** [reply]
> In any case, this should be generic enough that we can switch libraries internally, if needed. However, looking further, there's [a lot of boilerplate going on here](https://github.com/mcdo0486/pymeasure/blob/6d5de8eaba35bb1d1b4ab98b91859b29d6def551/pymeasure/instruments/omega/omega_registers.py).

Yea, my approach to programming functionality for that particular instrument is another different reason why I haven't looked to submit that branch as a PR. The [omega modbus interface](https://assets.omega.com/manuals/M5458.pdf) manual has a table with all the modbus register assignments including register type and address, so I moved the entire table over to a OMEGA_COMMANDS dictionary. It was one of the first instruments I worked on and I was trying something new.

I don't like the approach I took because I haven't used 99% of the recorded commands and I definitely wouldn't go through and test all of them to make sure they work. And I believe in testing instrument functionality as per the "untested" discussion.

I was bringing up my omega adapter simply to highlight minimalmodbus and its simplicity and ease of use.

**Reply by BenediktBurger** [reply]
I see the following difficulty:
We want to use adapters for the hardware and instruments for the protocol, such that you get the same experience with another adapter. (This is necessary for our testing etc as well). 

The bist way would be, to use a modbus instrument, because you could implement changes there and such change the library used. 

However, minimummodbus (and probably others) access directly the port (via pyserial), such that the proper place were an adapter. 
If we used an adapter, it would be easy to change (because the adapter could translate one parameter set to another), but if a part of the implementation is in the instrument (for example TC038D), modbus adapters are not possible anymore. 

Besides this basic challenge (how to connect), there is the additional challenge, how to transfer the parameters to the libraries methods, as our property creators generate a string only.
It is possible to encode the parameters (as I did), but it might get cumbersome if there are too many options to include in the string (for example float, int, long; how many decimals to multiply/divide...).

I see the fear to slowly growing into a copy of such a library.

In the end, the question is, how do we deal with protocols: implement ourselves (maybe partially) or import dependency and try to make it work in our setup (which will introduce other limitations to the library usage).

I guess it is less work to implement (a subset of) modbus than to make an external library work with our system.

**Reply by bilderbuchi** [reply]
Well argued, Benedikt, and I agree with your points. I agree that structure-wise, such a protocol should live in an Instrument, not an Adapter (also evidenced by the fact that minimalmodbus is just using pyserial in the background).

If we should create a `ModbusInstrument`, there is an interesting design aspect we have to think of: When we start implementing instrument classes (#416), we'll have e.g. a `PowerSupply` class that predefines a couple of standardized properties (`output_enabled`, `current_limit`,...) assuming some SCPI commands. How do we elegantly deal with the fact that an instrument can be classified along two different "axes" -- is it a power supply? does it communicate via modbus or scpi? The latter question influences the former command choices -- do we assume SCPI and if the instrument is not using that, just preserve the property names or override all command strings? Are mixins a good answer? Or better to use compostion? Or do we just define some API to be implemented via `abc`?
I realise this is more a discussion for #416, but it's something to keep in mind here.

**Reply by bilderbuchi** [reply]
The "frame-based" or "register-based" messaging of the TC038D is somewhat special already. One thing that could be a precursor to modbus integration, so we don't have to answer all questions in one go, could be to refactor the "frame-based messaging" bits into a (e.g.) `FrameBasedInstrument`, which would serve as the parent for the TC038D _and_ all other related instruments in the codebase (e.g. #239). 
This could already educate us about the trade-offs and effort involved and how the structure and instrument-side usage shakes out (e.g. a common and flexible command syntax to feed the necessary arguments via strings). 
As this is already pretty similar to Modbus, maybe after that we can see how the Omega temperature controller would be implemented in such a scheme.

What do you think?

**Reply by BenediktBurger** [reply]
I don't think, that `FrameBasedInstrument` is a good idea. Because the frame setup depends on the protocol used such that we need one Instrument for every protocol (I do not expect many different protocols used by man instruments).
I have to check, but I guess the TC038D is modbus, therefore we could just experiment with moving its code into a parent class.

**Reply by bilderbuchi** [reply]
> the frame setup depends on the protocol used

What I've seen so far is
* command-type
* address,
* length/type
* value (if setter)?

Does the way how this is assembled vary often?
The thing I've seen vary between instruments is if/which kind of checksumming is used.

Maybe the specifics can be isolated in a method like `insert_id()` for the channels?

**Reply by BenediktBurger** [reply]
Nice idea with some kind of `compose_message`.

**Comment by BenediktBurger** [comment]
I understand you and I don't like to repeat their work, but I also do not want to introduce another dependency which works differently from our code base.

Now we have to decide, whether and how we want to introduce it.

**Reply by bilderbuchi** [reply]
Let's not move too quickly, we don't have to decide anything now. It's good to see some demand, and what the needs could be. Let's wait to hear back from @sansanda first, in any case, maybe his device can already be implemented like the TC038D.

**Reply by BenediktBurger** [reply]
With "now", I meant as the next step, not "right now" :grin: 
Right now, we have more pressing matters to sort out.

**Comment by sansanda** [comment]
Hi All.

@bilderbuchi for the moment I've nothing to show with modbus communication. All I've done is reading some registers of an Eurotherm 2704 controller using pymodbus, but nothing in the framework of pymeasure. 

At first ModbusInstrument it seemed like a good idea but then what happen with GPIB instruments and with the Serial ones? I'm not sure.

Since modbus protocol itself can use serial or tcp/ip communication base on the protocol type: RTU, ASCII, TCP/IP, etc. perhaps it is not a good idea (I thought so) extending the serial adapter, instead we could create a new adapter which can use (wrapping) the VISA adapter in based on a parameter ("RTU" or "TCP/IP" for the moment). Internally, such adapter can construct the frame based on the protocol type.

In such aapter we also can use the TC038d functions for the CRC, write and read (perhaps with less options).

**Reply by BenediktBurger** [reply]
It's not a problem, to have a specific "parent" instrument class for modbus, as modbus is the protocol (how to compose a message), while Serial/GPIB/TCP is a hardware connection (which is dealt with in the adapter). In that case, you can use the VISAAdapter, SerialAdapter, and ProtocolAdapter in combination with the ModbusInstrument to get your messages to the device (or test).

So you think it sufficient to have a couple of methods for reading/writing registers?

**Reply by sansanda** [reply]
The new adapter should define the write single/multiple registers and so on based on the modbus protocol. The ideal situation is that such methods make use of an unique write/read methods (overriding in the same that is done in the TC038d).

**Reply by BenediktBurger** [reply]
Would it be sufficient to implement the `read_register` (3), `write_single_register` (6), and `write_registers` (16)?

Here is the [official protocol specification](https://modbus.org/docs/Modbus_Application_Protocol_V1_1b3.pdf).

I verified, the TC038D uses modbus.

**Reply by bilderbuchi** [reply]
Function code `03` can also read multiple registers, right? Because that is quite useful to get a lot of data in one request.
Thanks for the link to the spec, that is going to be useful.

I would imagine that our properties use tuples made up of function code (possibly made readable with an enum), starting address, length, value, etc as appropriate for the commands, and the `read()` dispatches to the correct specific method base on function code? 
Tuples could be comma-separated strings as we already did. Maybe we could investigate if we could manage to pass in/out bytearrays or some other structured object, although I suspect that could be challenging.

**Reply by bilderbuchi** [reply]
One "adventurous" avenue: 
We could *right now* make `minimalmodbus.Instrument` use one of our `SerialAdapter` instances if we are fine with using the private module-level [`_serialports`](https://github.com/pyhys/minimalmodbus/blob/185afffd2d65b585f69bc5e3f640270d2bd27444/minimalmodbus.py#L61), because that thing seems to be architected for connection reuse, anyway. 
This way, we could stay within our separation of protocol (`Instrument`) and connection (`Adapter`), but could make use of all the details already implemented in `minimalmodbus.Instrument` - - the specification is 50 pages, and the test suite is 6000 lines long, after all. :D
Possibly (unconfirmed) this would even be compatible with `VISAAdapter` (the `serial.Serial` API usage seems simple enough). Maybe the maintainer would take a pull request for enabling one to pass in an existing connection upon construction.

**Reply by BenediktBurger** [reply]
I like that idea: minimalmodbus is small, seems helpful and it could be relatively easy to make it work together with our code.

@sansanda when du you need an initial version to work? What is your plan?

EDIT: Note to me: It should be sufficient to replace minimalmodbus.Instrument's init and _communicate in order to make an Instrument compatible with our adapters.

**Reply by sansanda** [reply]
I like too, minimalbus is a good reference and has a lot of reusable code. 
My plan:

1. Make the modbus adapter in mode RTU and TCP/IP (as wrapper of VISA) able to communicate via VISA both Serial and Ethernet based on the connection string. I think this will be the most simple task, perhaps only is question of bypassing the connection string. I will see.
2. Test the adapter communication over a real instrument (our temperature controller for example) without implementing any function.
3. For the moment add to the adapter two functions for reading and writing single registers. 
4. Create a new instrument which will use the new adapter and code some simple properties.
5. The most tricky and challenging as @bilderbuchi points, try to integrate dynamic properties of the new instrument (f.e temperature_setpoint) with the underlying read/write methods and the different frame_formats of the new protocol.
6. Finally, if all works fine, complete the adapter with other functions for read/write multiple registers and aventually work with coils and other 1 bit operations.

**Reply by BenediktBurger** [reply]
Pleas do not use an adapter, but create a parent **instrument class**.

I thought about some
~~~python
class ModbusInstrument(modbus.Instrument, pymeasure.Instrument):
    def __init__(self, adapter, name="Modbus base instrument", address=1, **kwargs):
        pymeasure.Instrument.__init__(adapter, name, **kwargs)
        self.address = address  # modbus address
        # here is the necessary stuff of the modbus.Instrument.init

    def _communicate(self...):
        # adjust to use the adapter's read_bytes / write_bytes method, such that both VISAAdapter or SerialAdapter can be used.

    def ask(self, command):
        # separate command into the different parts (function code, values) and call the corresponding method, e.g. `read_register`
~~~

I'd say, that you can use either the modbus number as function code or some Letters (e.g. WR for write register, WC for write coil...).
Values can be comma separated in the string.

**Reply by sansanda** [reply]
I thought that you didn't want to create an instrument hierarchy. For me it's ok. Should then be implemented in instrument.py file? 
As I can see it seems that you want to introduce a dependency (minimalmodbus.Instrument). I that right?

**Reply by BenediktBurger** [reply]
> Should then be implemented in instrument.py file?

It should be in a different file.

> As I can see it seems that you want to introduce a dependency (minimalmodbus.Instrument). I that right?

Yes I do. That way we do not have to do everything from scratch. At least that is what we want to try now.

Let us try to create such an instrument, how much work it is, and how easy it works together with our infrastructure.

If that **preliminary** test is successful, we can define the right protocol to encode the strings sent from the property creators.

1. So the first test is, to create an instrument, which uses an adapter string (or adapter instance) and talks modbus with the device
2. Then we think how to connect the property creators (`read` and `write`, or `ask` methods) to the modbus methods
3. Implement these connections (at least for the most important). The good thing is, we always have the original modbus commands remaining in the code (`write_bits` etc.)

**Reply by sansanda** [reply]
Ok. Perfect. Working on it so.

**Reply by BenediktBurger** [reply]
> Function code 03 can also read multiple registers, right? Because that is quite useful to get a lot of data in one request.

Yes, it does, @bilderbuchi .

**Reply by bilderbuchi** [reply]
@bmoneke I'm wondering, instead of going multiple inheritance (`class ModbusInstrument(modbus.Instrument, pymeasure.Instrument)`), should we try to use composition first, (i.e. doing in `ModbusInstrument`: `self.modbus = modbus.Instrument(foo, bar)`) instead? That would reduce the potential for collision, not put stuff into the main instrument namespace that we don't want/need, and in general would be more flexible. 
What do you think?

**Reply by BenediktBurger** [reply]
At first it seemed very good to combine the methods, as such we get easy access to all the modbus functions without an additional "translator". That way an implementor might access easily the modbus functions with self. 

Yeah, thinking more about it, I found it difficult to make it work with the right super calls.

**Reply by bilderbuchi** [reply]
Often composition is preferable to inheritance, I think this here is such a situation.

**Reply by sansanda** [reply]
Yes. Perhaps inheritance is more convenient when we need polymorphism but I think is not our case. The risk of namespace collisions also counts.

**Reply by BenediktBurger** [reply]
I opened an [issue 92](https://github.com/pyhys/minimalmodbus/issues/92) with minimalmodbus regarding adding some kind of connection object.

Nevertheless, we need a `ModbusShim` that takes an adapter as input and offers `read`, `write` methods (which call the bytes versions of the adapter).
In order to simulate a serial object, the Shim has to implement `is_open`, `open()`, `port`, `reset_input_buffer()` `reset_output_buffer()`, `baudrate`, `write()`, `read()`, `flush()`, `close()`, `timeout`. (List is taken from the calls to the serial object, methods have parenthesis, attributes not)

So we would have something like
~~~python
class ModbusShim:
    def __init__(self, adapter):
        self.adapter = adapter

    def write(self, command):
        self.adapter.write_bytes(command)

    def read(self, count):
        return self.adapter.read_bytes(count)

    @property
    def baudrate(self):
        return self.adapter.connection.baudrate  # verify name with pyvisa
        # The protocolAdapter needs a large baudrate (larger 22000) to reduce test time to minimum

class ModbusInstrument(Instrument):
    def __init__(self, adapter, name="Modbus Instrument", address=1, **kwargs):
        super().__init__(adapter, name, **kwargs)
        if isinstance(adapter, ProtocolAdapter):
            adapter = name
        minimalmodbus._serialports[adapter] = ModbusShim(self.adapter)
        modbus = minimalmodbus.Instrument(adapter, address)

    def ask(self, command):  # (or split it into the read and write methods? -> discuss later)
        # separate the strings
        if a register should be read:
            return self.modbus.read_register(0x13, number_of_decimals=1)
~~~

**Comment by BenediktBurger** [comment]
Minimalmodbus is going to implement a fix, such that it it can accept a serial address string, which will make it easier to integrate into pymeasure:
- https://github.com/pyhys/minimalmodbus/issues/92

**Reply by BenediktBurger** [reply]
It is fixed. Now such a `minimalmodbus.Instrument` can accept any object, which can read and write from a connection.
It can be a serial object or a VISA Adapter (with some shimming, translating from `read_bytes` to `read` etc.).

**Reply by BenediktBurger** [reply]
My proposal how to implement it:
~~~python

class ModbusShim:
    """make Adapters appear like a `serial.Serial` object."""
    def __init__(self, adapter: Adapter):
        self.adapter = adapter

    def write(self, content: bytes):
        self.adapter.write_bytes(content)

class ModbusMixin:
    def __init__(self, **kwargs):
        super().__init__(**kwargs)
        self.modbus = minimalmodbus.Instrument(port = ModbusShim(self.adapter))


class ExampleModbusInstrument(ModbusMixin, Instrument):
    @property
    def temperature(self):
        """Control some temperature."""
        return self.modbus.read_register(...)

    @temperature.setter
    def temperature(self, content):
        self.modbus.write_register(content)
~~~

@bilderbuchi Any ideas on the general setup?

**Reply by mcdo0486** [reply]
That's great you initiated that change into the minimalmodbus library.

I think a mixin is the way to go. I'm guessing you're looking for feedback on the class layout, but I'll add a comment for the `ModbusMixin`.

There are four convenience functions to read/write  in minimalmodbus, `read_registers`/`write_registers` [uint16], `read_long`/`write_long` [default uint32], `read_float`/`write_float` [default IEEE 754 32bit] , and `read_string`/`write_string` [default 32 chars]. 

I'm wondering if `ModbusMixin` can setup some scaffolding to use the defaults with those functions in combination with `Instrument.control` to match common pymeasure driver implementation. However I think the `ModbusShim`.write() would break that? Ignoring the `ModbusShim` part, this is just a short mockup example:

```python
class ModbusShim:
    """make Adapters appear like a `serial.Serial` object."""
    def __init__(self, adapter: Adapter):
        self.adapter = adapter

    def write(self, content: bytes):
        self.adapter.write_bytes(content)

class ModbusMixin:

    def __init__(self, **kwargs):
        super().__init__(**kwargs)
        self.modbus = minimalmodbus.Instrument(port = ModbusShim(self.adapter))

   def write(self, command):
       split_com = command.split(':')
       com_type = split_com[0]
       register = split_com[1]
       value = split_com[2]
       if com_type == 'f':
            self.modbus.write_float(register, float(value))
       elif com_type ==  'r':
            self.modbus.write_register(register, int(value))

   def ask(self, command):
       split_com = command.split(':')
       com_type = split_com[0]
       register = split_com[1]
       if com_type == 'f':
            return self.modbus.read_float(register)
       elif com_type ==  'r':
            return self.modbus.read_register(register)

class ExampleModbusInstrument(ModbusMixin, Instrument):

    temperature = Instrument.control(("f:4023", "f:4023:%g", "Control the temp")

    status = Instrument.control(("r:2013", "r:2013:%d", "Control the status")

```

There is a lot of variety with modbus instruments, and if the minimalmodbus defaults for those functions don't work then a custom property could be used, or those methods can be implemented in the instrument driver.

Having a standard pattern for minimalmodbus that in the mixin that enables `Instrument.control` would be beneficial.

**Reply by BenediktBurger** [reply]
> I'm guessing you're looking for feedback on the class layout, but I'll add a comment for the 

 One point is class layout (mixin vs. inheritance), the other one, is how to implement reading registers.

In my current `hcp.TC038` implementation I use a setup similar to your proposal: The command string encodes the access type (read or write) and the register etc. The advantage is, that we can use `control`. The disadvantage is, that an implementor does not get the support of the IDE (showing the signature of the method) and readability is worse compared to `self.modbus.read_register(register=0x14)`.

For these reasons I opted for plain python properties in my example.

We could introduce a special form of `control`, which uses read/write registers, but that solution does not convince me, either.
~~~python
whatever = ModbusMixin.modbus_control(type="uint16", register=0x4213, doc="""Control some property.""")
# "uint16" could be done with enums as well.
~~~

> I think a mixin is the way to go.

Mixins have their downside as well, as you miss some methods. For example this mixin would not know anything about an Instruments methods (and the `self.adapter` appears magically). In this case, however, `Instrument` and the mixin do not touch, except in the `adapter`, therefore there should not be any problem.

**Reply by bilderbuchi** [reply]
> @bilderbuchi Any ideas on the general setup?

Sorry, I have to admit I lost the context from the discussion we had last year. It will be some time before I can pick it up again (and have other loose ends I want to tie down, first).
I assume that my comments from November are still relevant. 
One thing stuck out -- please try to keep in mind a future approach we'd choose for implementing #416 -- would be great if this were compatible.

**Reply by BenediktBurger** [reply]
The upstream code change triggered me finishing this open end and ideas on how to implement it.
At least I documented the rough idea. I'm fine with keeping it in the backlog, as there seems to be no urgency (my driver works and @sansanda did mention any questions in this regard).
I opened an issue (#949 ) to mark it as a todo.

**Comment by loicguilmard** [comment]
Hello,
I am currently working to write a pymodaq plugin for a cellkraft precision steam generator e-series with modbustcp.
What's the status or the hold on this mixin ?

**Reply by BenediktBurger** [reply]
Currently, we do not have the modbus support in pymeasure itself.

I don't know, whether @sansanda did anything usable.
So, no usable code.

For a simple, plain implementation, you could look at the HCP TC038D driver, which implements modbus by itself (I did not know, that it is modbus at that time).

You could copy the methods from there, if you want to.
Or you could go on and develop a pymeasure code for that according to ideas mentioned in this discussion.
